Basic RISC V support

0xMirasio opened the pull request:
I have basic riscv support with those

However capstone dont support RISC-V in his stable branch yet, so we must install capstone-5.0.0rc2 ( branch:next) 

Yet i added some gadget but there isnt big support for the moment because capstone isnt supporting some instruction yet . Cant test too much without capstone to advance on the riscv project.

0xMirasio commented:
Edit : forget to remove test_riscv.py file

SweetVishnya commented:
Thx! Supporting RISC-V is supper cool! Give me a few weeks to review this.

SweetVishnya commented:
Closes #129

SweetVishnya commented:
Btw, can you fix tests in CI and add some tests for RISC-V? Also, updating README.md and help messages with RISC-V support would be neat.

0xMirasio commented:
updated readme and CI should be okay. I'm working to add all the gadgets of jump (only a few are supported in my commits)

0xMirasio commented:
I added most of the 32bit encoded jump instruction for the riscvc set , should be Okay with most of the gadgets now. I'm working to add all the 64 bits encoded instructions from riscv64 but will take me some time.

0xMirasio commented:
now should be all good. added all gadgets for riscV 64bits

SweetVishnya commented:
Can you add binaries for both risv 32 and 64 bit? And run them in [test.sh](https://github.com/JonathanSalwan/ROPgadget/blob/master/test-suite-binaries/test.sh)?

SweetVishnya commented on ropgadget/loaders/elf.py:
```diff
@@ -340,6 +341,8 @@ def getArch(self):
             return CS_ARCH_PPC
         elif self.__ElfHeader.e_machine == ELFFlags.EM_SPARCv8p:
             return CS_ARCH_SPARC
+        elif self.__ElfHeader.e_machine == ELFFlags.EM_RISCV:
```
So, there are no separate flags for riscv 32 and 64?

0xMirasio replied:
yes

SweetVishnya commented on ropgadget/gadgets.py:
```diff
@@ -338,6 +377,13 @@ def addSYSGadgets(self, section):
                 arch_mode = CS_MODE_ARM
+        elif arch == CS_ARCH_RISCV:
+            gadgets = [
+                     [b"\x73\x00\x00\x00", 4, 2], #ecall                
```
does it depend on endianness?

0xMirasio replied:
yes, will add it

0xMirasio commented:
i Added 32bits sample , but i don't have time to add the support for it. I added RISCV Compressed set and 64bits. You can still use the tool on 32Bits binary because it contain a lot of compressed instructions.

SweetVishnya commented on test-suite-binaries/test.sh:
```diff
@@ -61,5 +61,9 @@ echo "RUN elf-ARM64-bash --depth 5" | tee -a  ./test_output
+echo "RUN elf-Linux-RISCV_64 --depth 8" | tee -a  ./test_output
+$RUN --binary ./elf-PPC64-bash --depth 5 1>> ./test_output
+echo "RUN elf-Linux-RISCV_32 --depth 8" | tee -a  ./test_output
+$RUN --binary ./elf-PPC64-bash --depth 5 1>> ./test_output
```
It seems that you forgot to replace ppc binary with risv in test.sh

0xMirasio replied:
fix

SweetVishnya commented:
> i Added 32bits sample , but i don't have time to add the support for it. I added RISCV Compressed set and 64bits. You can still use the tool on 32Bits binary because it contain a lot of compressed instructions.

No problem, supporting riscv 64 and compressed is still awesome!)

SweetVishnya commented on README.md:
```diff
@@ -67,6 +67,8 @@ Usage
       - PowerPC
       - Sparc
+      - RISC-V 64
+      - RISC-V Compressed
```
Can you add the same in [ropgadget/args.py](https://github.com/JonathanSalwan/ROPgadget/blob/master/ropgadget/args.py#L48)?

0xMirasio replied:
Done.

SweetVishnya commented on ropgadget/loaders/elf.py:
```diff
@@ -26,6 +26,7 @@ class ELFFlags(object):
     EM_PowerPC  = 0x14
     EM_PPC64    = 0x15
     EM_ARM64    = 0xb7
+    EM_RISCV    = 0Xf3
```
There is also [raw](https://github.com/JonathanSalwan/ROPgadget/blob/master/ropgadget/loaders/raw.py#L35) format loader.

0xMirasio replied:
added support for riscv 64b | compressed. will not add 32bit yet
